plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts

```diff
@@ -657,29 +637,135 @@ describe('KubernetesProxy', () => {
+    });
+
     const response = await requestPromise;
 
     expect(response.status).toEqual(200);
   });
+
+  describe('WebSocket', () => {
+    const proxyPath = '/proxy';
+    const proxyPort = 9000;
+    const wsPath = '/ws';
+    const wsPort = 9999;
+
+    let wsServer: WebSocketServer;
+    let expressApp: express.Express;
+    let expressServer: Server;
+
+    const eventPromiseFactory = (
+      ws: WebSocket,
+      event: 'connection' | 'open' | 'close' | 'error' | 'message',
+    ) => new Promise(resolve => ws.once(event, x => resolve(x?.toString())));
+
+    beforeAll(async () => {
+      await new Promise(resolve => {
+        expressApp = express().use(
+          Router()
+            .use(proxyPath, proxy.createRequestHandler({ permissionApi }))
+            .use(errorHandler()),
+        );
+
+        expressServer = expressApp.listen(proxyPort, '0.0.0.0', () => {
+          resolve(null);
+        });
+      });
+
+      wsServer = new WebSocketServer({ port: wsPort, path: wsPath });
+
+      wsServer.on('connection', (ws: WebSocket) => {
+        // send immediatly a feedback to the incoming connection
+        ws.send('connected');
+
+        // Echo message handling
+        ws.on('message', (message: string) => {
+          ws.send(message);
+        });
+      });
+
+      wsServer.on('error', console.error);
+    });
+
+    afterAll(() => {
+      wsServer.close();
+      expressServer.close();
+    });
+
+    // eslint-disable-next-line jest/no-done-callback
+    it('should proxy websocket connections', async () => {
+      clusterSupplier.getClusters.mockResolvedValue([
+        {
+          name: 'local',
+          url: `http://localhost:${wsPort}`,
+          serviceAccountToken: '',
+          authProvider: 'serviceAccount',
+        },
+      ] as ClusterDetails[]);
+
+      authTranslator.decorateClusterDetailsWithAuth.mockResolvedValue({
+        name: 'local',
+        url: `http://localhost:${wsPort}`,
+        serviceAccountToken: '',
+        authProvider: 'serviceAccount',
+      } as ClusterDetails);
+
+      const wsProxyAddress = `ws://localhost:${proxyPort}${proxyPath}${wsPath}`;
+      const wsAddress = `ws://localhost:${wsPort}${wsPath}`;
+
+      // Let this request through so it reaches the express router above
+      worker.use(
+        rest.all(wsAddress.replace('ws', 'http'), (req: any) =>
+          req.passthrough(),
+        ),
+        rest.all(wsProxyAddress.replace('ws', 'http'), (req: any) =>
+          req.passthrough(),
+        ),
+      );
+
+      // Prepopulate the proxy so the WebSocket upgrade can happen, result doesn't actually matter
+      const result = await fetch(wsProxyAddress.replace('ws', 'http'));
+      expect(result.ok).toBeFalsy();
+
+      const webSocket = new WebSocket(wsProxyAddress);
+
+      const connectMessagePromise = eventPromiseFactory(webSocket, 'message');
+
+      const openPromise = eventPromiseFactory(webSocket, 'open');
+      await openPromise;
```

Comment:
Little nit, not blocking -- I think you can await it without giving a name? Just to save a line really
```suggestion
      await eventPromiseFactory(webSocket, 'open');
```

Reply:
I also think it should work, probably had it like this because I was asserting something about the resolved value before creating the `eventPromiseFactory` factory, will use the suggestion and see if it still passes.